[HarmonyHub] Timeout-Exceptions

<!-- Provide a general summary of the issue in the *Title* above -->
Many Timeout-Exceptions - the reason is quite unclear.

## Expected Behavior
Better explanations where and why the timeout occure. 
Much better - not timeout exception at all.

## Current Behavior

22:07:19.216 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'harmonyhub.things'
22:07:24.226 [ERROR] [ome.core.thing.internal.ThingManager] - Exception occurred while calling thing updated at ThingHandler 'org.openhab.binding.harmonyhub.handler.HarmonyHubHandler@62cfd6cd': null
java.util.concurrent.TimeoutException
	at java.util.concurrent.FutureTask.get(FutureTask.java:205)[:1.8.0_144]
	at org.eclipse.smarthome.core.common.SafeMethodCaller.callAsynchronous(SafeMethodCaller.java:194)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:83)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:67)
	at org.eclipse.smarthome.core.thing.internal.ThingManager.thingUpdated(ThingManager.java:524)
	at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.notifyTrackers(ThingRegistryImpl.java:221)
	at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.notifyListenersAboutUpdatedElement(ThingRegistryImpl.java:144)
	at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.notifyListenersAboutUpdatedElement(ThingRegistryImpl.java:1)
	at org.eclipse.smarthome.core.common.registry.AbstractRegistry.updated(AbstractRegistry.java:177)[98:org.eclipse.smarthome.core:0.9.0.b5]
	at org.eclipse.smarthome.core.common.registry.AbstractProvider.notifyListeners(AbstractProvider.java:57)[98:org.eclipse.smarthome.core:0.9.0.b5]
	at org.eclipse.smarthome.core.common.registry.AbstractProvider.notifyListenersAboutUpdatedElement(AbstractProvider.java:82)[98:org.eclipse.smarthome.core:0.9.0.b5]
	at org.eclipse.smarthome.model.thing.internal.GenericThingProvider.lambda$18(GenericThingProvider.java:978)[133:org.eclipse.smarthome.model.thing:0.9.0.b5]
	at java.util.ArrayList.forEach(ArrayList.java:1249)[:1.8.0_144]
	at org.eclipse.smarthome.model.thing.internal.GenericThingProvider.createThingsFromModelForThingHandlerFactory(GenericThingProvider.java:987)[133:org.eclipse.smarthome.model.thing:0.9.0.b5]
	at org.eclipse.smarthome.model.thing.internal.GenericThingProvider.access$1(GenericThingProvider.java:950)[133:org.eclipse.smarthome.model.thing:0.9.0.b5]
	at org.eclipse.smarthome.model.thing.internal.GenericThingProvider$1.lambda$0(GenericThingProvider.java:234)[133:org.eclipse.smarthome.model.thing:0.9.0.b5]
	at java.util.concurrent.ConcurrentHashMap$KeySetView.forEach(ConcurrentHashMap.java:4649)[:1.8.0_144]
	at org.eclipse.smarthome.model.thing.internal.GenericThingProvider$1.apply(GenericThingProvider.java:236)[133:org.eclipse.smarthome.model.thing:0.9.0.b5]
	at org.eclipse.smarthome.model.thing.internal.GenericThingProvider$1.apply(GenericThingProvider.java:1)[133:org.eclipse.smarthome.model.thing:0.9.0.b5]
	at org.eclipse.smarthome.config.core.BundleProcessorVetoManager.applyActionFor(BundleProcessorVetoManager.java:124)
	at org.eclipse.smarthome.model.thing.internal.GenericThingProvider.lambda$3(GenericThingProvider.java:304)[133:org.eclipse.smarthome.model.thing:0.9.0.b5]
	at java.lang.Iterable.forEach(Iterable.java:75)[:1.8.0_144]
	at org.eclipse.smarthome.model.thing.internal.GenericThingProvider.createThingsFromModel(GenericThingProvider.java:306)[133:org.eclipse.smarthome.model.thing:0.9.0.b5]
	at org.eclipse.smarthome.model.thing.internal.GenericThingProvider.modelChanged(GenericThingProvider.java:820)[133:org.eclipse.smarthome.model.thing:0.9.0.b5]
	at org.eclipse.smarthome.model.core.internal.ModelRepositoryImpl.notifyListeners(ModelRepositoryImpl.java:286)[122:org.eclipse.smarthome.model.core:0.9.0.b5]
	at org.eclipse.smarthome.model.core.internal.ModelRepositoryImpl.addOrRefreshModel(ModelRepositoryImpl.java:120)[122:org.eclipse.smarthome.model.core:0.9.0.b5]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver.checkFile(FolderObserver.java:234)[122:org.eclipse.smarthome.model.core:0.9.0.b5]
	at org.eclipse.smarthome.model.core.internal.folder.FolderObserver.processWatchEvent(FolderObserver.java:297)[122:org.eclipse.smarthome.model.core:0.9.0.b5]
	at org.eclipse.smarthome.core.service.WatchQueueReader.run(WatchQueueReader.java:206)[98:org.eclipse.smarthome.core:0.9.0.b5]
	at java.lang.Thread.run(Thread.java:748)[:1.8.0_144]

## Possible Solution
increase default Timeout - but - my tests suggest no impact on this error. 

## Steps to Reproduce (for Bugs)
Just use HarmonyHub-Binding. No special preparations needed.

## Context
The issue is minor. The effect on functionality is not given. It just explodes the logs

## Your Environment
* Version used: (e.g., openhab and addon versions)
openhab 2.1

* Environment name and version (e.g. Chrome 39, node.js 5.4, Java 8, ...):
java version "1.8.0_144"
Java(TM) SE Runtime Environment (build 1.8.0_144-b01)
Java HotSpot(TM) 64-Bit Server VM (build 25.144-b01, mixed mode)

* Operating System and version (desktop or mobile, Windows 10, Raspbian Jessie, ...):
➜  services git:(master) ✗ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.3 LTS
Release:	16.04
Codename:	xenial

--- a/addons/binding/org.openhab.binding.harmonyhub/META-INF/MANIFEST.MF
+++ b/addons/binding/org.openhab.binding.harmonyhub/META-INF/MANIFEST.MF
@@ -34,6 +34,7 @@ Import-Package:
  org.eclipse.jdt.annotation;resolution:=optional,
  org.eclipse.smarthome.config.core,
  org.eclipse.smarthome.config.discovery,
+ org.eclipse.smarthome.core.cache,
  org.eclipse.smarthome.core.common,
  org.eclipse.smarthome.core.library.types,
  org.eclipse.smarthome.core.thing,

--- a/addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyDeviceHandler.java
+++ b/addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyDeviceHandler.java
@@ -144,7 +144,11 @@ public class HarmonyDeviceHandler extends BaseThingHandler {
         if (bridgeStatus == ThingStatus.ONLINE && getThing().getStatus() != ThingStatus.ONLINE) {
             bridge = (HarmonyHubHandler) getBridge().getHandler();
             updateStatus(ThingStatus.ONLINE);
-            bridge.getConfigFuture().thenAcceptAsync(this::updateChannel, scheduler);
+            bridge.getConfigFuture().thenAcceptAsync(this::updateChannel, scheduler).exceptionally(e -> {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "Getting config failed: " + e.getMessage());
+                return null;
+            });
         } else if (bridgeStatus != ThingStatus.ONLINE) {
             updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
         }

--- a/addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyDeviceHandler.java
+++ b/addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyDeviceHandler.java
@@ -144,7 +144,7 @@ public class HarmonyDeviceHandler extends BaseThingHandler {
         if (bridgeStatus == ThingStatus.ONLINE && getThing().getStatus() != ThingStatus.ONLINE) {
             bridge = (HarmonyHubHandler) getBridge().getHandler();
             updateStatus(ThingStatus.ONLINE);
-            updateChannel();
+            bridge.getConfigFuture().thenAcceptAsync(this::updateChannel, scheduler);
         } else if (bridgeStatus != ThingStatus.ONLINE) {
             updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
         }
@@ -153,11 +153,10 @@ public class HarmonyDeviceHandler extends BaseThingHandler {
     /**
      * Updates our channel with the available buttons as option states
      */
-    private void updateChannel() {
+    private void updateChannel(HarmonyConfig config) {
         try {
             logger.debug("updateChannel for device {}", logName);
 
-            HarmonyConfig config = bridge.getCachedConfig();
             if (config == null) {
                 logger.debug("updateChannel: could not get config from bridge {}", logName);
                 return;

--- a/addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyHubHandler.java
+++ b/addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyHubHandler.java
@@ -13,10 +13,10 @@ import static org.openhab.binding.harmonyhub.HarmonyHubBindingConstants.HARMONY_
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.Comparator;
-import java.util.Date;
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Set;
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
@@ -25,6 +25,7 @@ import java.util.concurrent.TimeUnit;
 
 import org.apache.commons.lang.StringUtils;
 import org.eclipse.smarthome.config.core.Configuration;
+import org.eclipse.smarthome.core.cache.ExpiringCacheAsync;
 import org.eclipse.smarthome.core.library.types.DecimalType;
 import org.eclipse.smarthome.core.library.types.StringType;
 import org.eclipse.smarthome.core.thing.Bridge;
@@ -68,6 +69,16 @@ public class HarmonyHubHandler extends BaseBridgeHandler implements HarmonyHubLi
 
     public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Collections.singleton(HARMONY_HUB_THING_TYPE);
 
+    private static final Comparator<Activity> ACTIVITY_COMPERATOR = new Comparator<Activity>() {
+        @Override
+        public int compare(Activity a1, Activity a2) {
+            // if the order value is null we want it to be at the start of the list
+            int o1 = a1.getActivityOrder() == null ? -1 : a1.getActivityOrder().intValue();
+            int o2 = a2.getActivityOrder() == null ? -1 : a2.getActivityOrder().intValue();
+            return (o1 < o2) ? -1 : (o1 == o2) ? 0 : 1;
+        }
+    };
+
     // one minute should be plenty short, but not overwhelm the hub with requests
     private static final long CONFIG_CACHE_TIME = 60 * 1000;
 
@@ -81,9 +92,7 @@ public class HarmonyHubHandler extends BaseBridgeHandler implements HarmonyHubLi
 
     private HarmonyClient client;
 
-    private HarmonyConfig cachedConfig;
-
-    private Date cacheConfigExpireDate;
+    private ExpiringCacheAsync<HarmonyConfig> configCache = new ExpiringCacheAsync<>(CONFIG_CACHE_TIME);
 
     private HarmonyHubHandlerFactory factory;
 
@@ -225,7 +234,11 @@ public class HarmonyHubHandler extends BaseBridgeHandler implements HarmonyHubLi
                 }
             }, heartBeatInterval, heartBeatInterval, TimeUnit.SECONDS);
             updateStatus(ThingStatus.ONLINE);
-            buildChannel();
+            getConfigFuture().thenAcceptAsync(harmonyConfig -> buildChannel(harmonyConfig), scheduler)
+                    .exceptionally(e -> {
+                        setOfflineAndReconnect("Getting config failed: " + e.getMessage());
+                        return null;
+                    });
         } catch (Exception e) {
             logger.debug("Could not connect to HarmonyHub at {}", host, e);
             setOfflineAndReconnect("Could not connect: " + e.getMessage());
@@ -275,14 +288,15 @@ public class HarmonyHubHandler extends BaseBridgeHandler implements HarmonyHubLi
             case HUB_IS_TURNING_OFF:
                 // hub is turning off is received for current activity, we will translate it into activity starting
                 // trigger of power-off activity (with ID=-1)
-                HarmonyConfig config = getCachedConfig();
-                if (config != null) {
-                    Activity powerOff = config.getActivityById(-1);
-                    if (powerOff != null) {
-                        triggerChannel(HarmonyHubBindingConstants.CHANNEL_ACTIVITY_STARTING_TRIGGER,
-                                getEventName(powerOff));
+                getConfigFuture().thenAccept(config -> {
+                    if (config != null) {
+                        Activity powerOff = config.getActivityById(-1);
+                        if (powerOff != null) {
+                            triggerChannel(HarmonyHubBindingConstants.CHANNEL_ACTIVITY_STARTING_TRIGGER,
+                                    getEventName(powerOff));
+                        }
                     }
-                }
+                });
                 break;
             default:
                 break;
@@ -293,20 +307,11 @@ public class HarmonyHubHandler extends BaseBridgeHandler implements HarmonyHubLi
         return activity.getLabel().replaceAll("[^A-Za-z0-9]", "_");
     }
 
-    private void buildChannel() {
+    private void buildChannel(HarmonyConfig config) {
         try {
-            HarmonyConfig config = getCachedConfig();
             List<Activity> activities = config.getActivities();
             // sort our activities in order
-            Collections.sort(activities, new Comparator<Activity>() {
-                @Override
-                public int compare(Activity a1, Activity a2) {
-                    // if the order value is null we want it to be at the start of the list
-                    int o1 = a1.getActivityOrder() == null ? -1 : a1.getActivityOrder().intValue();
-                    int o2 = a2.getActivityOrder() == null ? -1 : a2.getActivityOrder().intValue();
-                    return (o1 < o2) ? -1 : (o1 == o2) ? 0 : 1;
-                }
-            });
+            Collections.sort(activities, ACTIVITY_COMPERATOR);
 
             // add our activities as channel state options
             List<StateOption> states = new LinkedList<StateOption>();
@@ -348,15 +353,6 @@ public class HarmonyHubHandler extends BaseBridgeHandler implements HarmonyHubLi
     }
 
     /**
-     * Returns an instance of our connected HarmonyClient
-     *
-     * @return
-     */
-    public HarmonyClient getClient() {
-        return client;
-    }
-
-    /**
      * Sends a button press to a device
      *
      * @param device
@@ -388,23 +384,25 @@ public class HarmonyHubHandler extends BaseBridgeHandler implements HarmonyHubLi
         }
     }
 
-    /**
-     * Cache a call to get the harmony config so we don't overwhelm it with requests.
-     *
-     * @return
-     */
-    public synchronized HarmonyConfig getCachedConfig() {
-        if (client == null) {
-            return null;
-        }
-
-        Date now = new Date();
-        if (cachedConfig == null || cacheConfigExpireDate == null || now.after(cacheConfigExpireDate)) {
-            logger.debug("Refreshing conf cache");
-            cachedConfig = client.getConfig();
-            cacheConfigExpireDate = new Date(System.currentTimeMillis() + CONFIG_CACHE_TIME);
-        }
-        return cachedConfig;
+    public synchronized CompletableFuture<HarmonyConfig> getConfigFuture() {
+        return configCache.getValue(() -> {
+            return new CompletableFuture<HarmonyConfig>() {
+                {
+                    if (client == null) {
+                        logger.debug("Could not get config because client is null");
+                        complete(null);
+                    } else {
+                        try {
+                            logger.debug("Getting config from client");
+                            complete(client.getConfig());
+                        } catch (Exception e) {
+                            logger.debug("Could not get config from client", e);
+                            completeExceptionally(e);
+                        }
+                    }
+                }
+            };
+        });
     }
 
     /**

--- a/addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyHubHandler.java
+++ b/addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyHubHandler.java
@@ -22,6 +22,7 @@ import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
 
 import org.apache.commons.lang.StringUtils;
 import org.eclipse.smarthome.config.core.Configuration;
@@ -69,18 +70,11 @@ public class HarmonyHubHandler extends BaseBridgeHandler implements HarmonyHubLi
 
     public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Collections.singleton(HARMONY_HUB_THING_TYPE);
 
-    private static final Comparator<Activity> ACTIVITY_COMPERATOR = new Comparator<Activity>() {
-        @Override
-        public int compare(Activity a1, Activity a2) {
-            // if the order value is null we want it to be at the start of the list
-            int o1 = a1.getActivityOrder() == null ? -1 : a1.getActivityOrder().intValue();
-            int o2 = a2.getActivityOrder() == null ? -1 : a2.getActivityOrder().intValue();
-            return (o1 < o2) ? -1 : (o1 == o2) ? 0 : 1;
-        }
-    };
+    private static final Comparator<Activity> ACTIVITY_COMPERATOR = Comparator.comparing(Activity::getActivityOrder,
+            Comparator.nullsFirst(Integer::compareTo));
 
     // one minute should be plenty short, but not overwhelm the hub with requests
-    private static final long CONFIG_CACHE_TIME = 60 * 1000;
+    private static final long CONFIG_CACHE_TIME = TimeUnit.MINUTES.toMillis(1);
 
     private static final int RETRY_TIME = 60;
 
@@ -129,7 +123,7 @@ public class HarmonyHubHandler extends BaseBridgeHandler implements HarmonyHubLi
                 logger.error("Could not start activity", e);
             }
         } else {
-            logger.warn("Command {}: Not a acceptable type (String or Decimal), ignorning", command);
+            logger.warn("Command {}: Not an acceptable type (String or Decimal), ignoring", command);
         }
     }
 
@@ -296,6 +290,9 @@ public class HarmonyHubHandler extends BaseBridgeHandler implements HarmonyHubLi
                                     getEventName(powerOff));
                         }
                     }
+                }).exceptionally(e -> {
+                    setOfflineAndReconnect("Getting config failed: " + e.getMessage());
+                    return null;
                 });
                 break;
             default:
@@ -384,24 +381,22 @@ public class HarmonyHubHandler extends BaseBridgeHandler implements HarmonyHubLi
         }
     }
 
-    public synchronized CompletableFuture<HarmonyConfig> getConfigFuture() {
+    public CompletableFuture<HarmonyConfig> getConfigFuture() {
+        Supplier<HarmonyConfig> configSupplier = () -> {
+            if (client == null) {
+                throw new IllegalStateException("Client is null");
+            }
+            try {
+                logger.debug("Getting config from client");
+                return client.getConfig();
+            } catch (Exception e) {
+                logger.debug("Could not get config from client", e);
+                throw e;
+            }
+        };
+
         return configCache.getValue(() -> {
-            return new CompletableFuture<HarmonyConfig>() {
-                {
-                    if (client == null) {
-                        logger.debug("Could not get config because client is null");
-                        complete(null);
-                    } else {
-                        try {
-                            logger.debug("Getting config from client");
-                            complete(client.getConfig());
-                        } catch (Exception e) {
-                            logger.debug("Could not get config from client", e);
-                            completeExceptionally(e);
-                        }
-                    }
-                }
-            };
+            return CompletableFuture.supplyAsync(configSupplier, scheduler);
         });
     }
 

--- a/addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/internal/discovery/HarmonyDeviceDiscoveryService.java
+++ b/addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/internal/discovery/HarmonyDeviceDiscoveryService.java
@@ -75,17 +75,20 @@ public class HarmonyDeviceDiscoveryService extends AbstractDiscoveryService impl
      * Discovers devices connected to a hub
      */
     private void discoverDevices() {
-        if (bridge.getClient() == null) {
-            logger.debug("Harmony client not connected, scanning postponed.");
+        if (bridge.getThing().getStatus() != ThingStatus.ONLINE) {
+            logger.debug("Harmony Hub not online, scanning postponed");
             return;
         }
         logger.debug("getting devices on {}", bridge.getThing().getUID().getId());
-        HarmonyConfig config = null;
-
-        try {
-            config = bridge.getCachedConfig();
-        } catch (Exception e) {
+        bridge.getConfigFuture().thenAccept(this::addDiscoveryResults).exceptionally(e -> {
             logger.debug("Could not get harmony config for discovery, skipping");
+            return null;
+        });
+    }
+
+    private void addDiscoveryResults(HarmonyConfig config) {
+        if (config == null) {
+            logger.debug("addDiscoveryResults: skipping null config");
             return;
         }
 

